specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1082 @@
+    // This functions activates the subnet and from there on other
+    // subnets in the system are allowed to interact with it and the
+    // subnet can start commtting its checkpoints.
+    // (methodNum = 2)
+    Register()
+    
+    // AddCollateral expects as source the address of the subnet actor
+    // for which the collateral wants to be added.  Its `value` should
+    // include the amount of collateral to be added for the subnet.
+    // (methodNum = 3)
+    AddCollateral()
+    
+    // ReleaseCollateral expects as source the address of the subnet actor
+    // for which the collateral should be released. It triggers a message
+    // to the subnet actor returning the corresponding collateral. 
+    // (methodNum = 4)
+    ReleaseCollateral(value abi.TokenAmount)
+    
+    // Kill expects as source the address of the subnet actor to be killed.
+    // This function can only be executed if no collateral or circulating
+    // supply is left for the subnet (i.e. balance = 0).
+    // (methodNum = 5)
+    Kill()
+    
+    // Kill expects as source the address of the subnet actor for which the
+    // checkpoint is being committed. The function performs some basic checks
+    // to ensure that checkpoint is valid and it persist it in the SCA state.
+    // (methodNum = 6)
+    CommitChildCheckpoint(ch Checkpoint)
+    
+    // Fund can be called by any user in a subnet and it injects 
+    // the `value` of native tokens included in the message to the source's address in
+    // the child subnet given as argument.
+    // (methodNum = 7)
+    Fund(address.SubnetID)
+    
+    // Release can be called by any user in a subnet to release the amount 
+    // of native tokens included in `value` from its own address in the
+    // subnet to the address in the parent. 
+    // (methodNum = 8)
+    Release()
+    
+    // SendCross can be called by any user in the subnet to send 
+    // an arbitrary cross-message to any other subnet in
+    // the hierarchy.
+    // (methodNum = 9)
+    SendCross(msg CrossMsgParams)
+    
+    // ApplyMessage can only be called as an `ImplicitMessage` by
+    // the `SystemActor` and is used to perform the execution of a 
+    // of cross-net messages in the subnet. 
+    // (methodNum = 10)
+    ApplyMessage(msg CrossMsgParams)
+    
+    // InitAtomicExec can be called by users to initiate an
+    // atomic execution with some other subnet.
+    // (methodNum = 11)
+    InitAtomicExec(params AtomicExecParams)
+    
+    // SubmitAtomicExec has to be called by all participants in an
+    // atomic execution to submit their results and trigger the 
+    // propagation of the output (or the abortion) of the execution
+    // to the corresponding subnets.
+    // (methodNum = 12)
+    SubmitAtomicExec(submit SubmitExecParams)
+    
+    // WIP: Not implemented yet.
+    // (methodNum = 13)
+    ReportMisbehavior()
+    
+    // WIP: Not implemented yet.
+    // (methodNum = 14)
+    Save()
```

Comment:
Should we specify the interface anyway? That's usually how things are supposed to go 😄

Reply:
Yes! Again, I left it as homework for "future-me" and completely missed it. I will add a proposal (even if we end up changing it). Thanks!